builtin/providers/azurerm/resource_arm_app_gateway.go

```diff
@@ -0,0 +1,1667 @@
+				ID: &v,
+			}
+		}
+
+		name := data["name"].(string)
+		frontEndConfig := network.ApplicationGatewayFrontendIPConfiguration{
+			Name: &name,
+			ApplicationGatewayFrontendIPConfigurationPropertiesFormat: &properties,
+		}
+
+		frontEndConfigs = append(frontEndConfigs, frontEndConfig)
+	}
+
+	return &frontEndConfigs
+}
+
+func expandAppGatewayBackendAddressPools(d *schema.ResourceData) *[]network.ApplicationGatewayBackendAddressPool {
+	configs := d.Get("backend_address_pool").([]interface{})
+	backendPools := make([]network.ApplicationGatewayBackendAddressPool, 0, len(configs))
+
+	for _, configRaw := range configs {
+		data := configRaw.(map[string]interface{})
+
+		backendAddresses := []network.ApplicationGatewayBackendAddress{}
+
+		for _, rawIP := range data["ip_address_list"].([]interface{}) {
+			ip := rawIP.(string)
+			backendAddresses = append(backendAddresses, network.ApplicationGatewayBackendAddress{IPAddress: &ip})
+		}
+
+		for _, rawFQDN := range data["fqdn_list"].([]interface{}) {
+			fqdn := rawFQDN.(string)
+			backendAddresses = append(backendAddresses, network.ApplicationGatewayBackendAddress{Fqdn: &fqdn})
+		}
+
+		name := data["name"].(string)
+		pool := network.ApplicationGatewayBackendAddressPool{
+			Name: &name,
+			ApplicationGatewayBackendAddressPoolPropertiesFormat: &network.ApplicationGatewayBackendAddressPoolPropertiesFormat{
+				BackendAddresses: &backendAddresses,
+			},
+		}
+
+		backendPools = append(backendPools, pool)
+	}
+
+	return &backendPools
+}
+
+func expandAppGatewayBackendHTTPSettings(d *schema.ResourceData, gatewayID string) *[]network.ApplicationGatewayBackendHTTPSettings {
+	configs := d.Get("backend_http_settings").([]interface{})
+	backendSettings := make([]network.ApplicationGatewayBackendHTTPSettings, 0, len(configs))
+
+	for _, configRaw := range configs {
+		data := configRaw.(map[string]interface{})
+
+		name := data["name"].(string)
+		port := int32(data["port"].(int))
+		protocol := data["protocol"].(string)
+		cookieBasedAffinity := data["cookie_based_affinity"].(string)
+		requestTimeout := int32(data["request_timeout"].(int))
+
+		setting := network.ApplicationGatewayBackendHTTPSettings{
+			Name: &name,
+			ApplicationGatewayBackendHTTPSettingsPropertiesFormat: &network.ApplicationGatewayBackendHTTPSettingsPropertiesFormat{
+				Port:                &port,
+				Protocol:            network.ApplicationGatewayProtocol(protocol),
+				CookieBasedAffinity: network.ApplicationGatewayCookieBasedAffinity(cookieBasedAffinity),
+				RequestTimeout:      &requestTimeout,
+			},
+		}
+
+		if data["authentication_certificate"] != nil {
+			authCerts := data["authentication_certificate"].([]interface{})
+			authCertSubResources := make([]network.SubResource, 0, len(authCerts))
+
+			for _, rawAuthCert := range authCerts {
+				authCert := rawAuthCert.(map[string]interface{})
+				authCertID := fmt.Sprintf("%s/authenticationCertificates/%s", gatewayID, authCert["name"])
```

Comment:
given there's no guarantee that these URL's won't change in the future, would it make more sense to pass in the Application Certificate ID, rather than the name?